src/riak_kv_ts_util.erl

```diff
@@ -97,15 +110,111 @@ build_sql_record(insert, SQL, _Cover) ->
             V = proplists:get_value(values, SQL),
             {ok, #riak_sql_insert_v1{'INSERT'   = T,
                                      fields     = F,
-                                     values     = V,
+                                     values     = convert_insert_timestamps(Mod, F, V),
                                      helper_mod = Mod
                                     }};
         false ->
             {error, <<"Must provide exactly one table name">>}
     end;
-build_sql_record(show_tables, _SQL, _Cover) ->
+build_sql_record_int(show_tables, _SQL, _Cover) ->
     {ok, #riak_sql_show_tables_v1{}}.
 
+convert_where_timestamps(_Mod, []) ->
+    [];
+convert_where_timestamps(Mod, Where) ->
+    [replace_ast_timestamps(Mod, hd(Where))].
+
+replace_ast_timestamps(Mod, {Op, Item1, Item2}) when is_tuple(Item1) andalso is_tuple(Item2) ->
+    {Op, replace_ast_timestamps(Mod, Item1), replace_ast_timestamps(Mod, Item2)};
+replace_ast_timestamps(Mod, {Op, FieldName, {binary, Value}}) ->
+    {Op, FieldName, maybe_convert_to_epoch(catch Mod:get_field_type([FieldName]), Value, Op)};
+replace_ast_timestamps(_Mod, {Op, Item1, Item2}) ->
+    {Op, Item1, Item2}.
+
+do_epoch(String, CompleteFun, Exponent) ->
+    Normal = timestamp_to_normalized(String),
+    case jam:is_valid(Normal) of
+        false ->
+            throw(<<"Invalid date/time string">>);
+        true ->
+            ok
+    end,
+    Epoch = case jam:is_complete(Normal) of
+                true ->
+                    jam:to_epoch(Normal, Exponent);
+                false ->
+                    jam:to_epoch(CompleteFun(Normal), Exponent)
+            end,
+    case is_atom(Epoch) of
+        true ->
+            %% jam returns atoms if processing gives an error
+            throw(Epoch);
+        false ->
+            Epoch
+    end.
+
+maybe_convert_to_epoch({'EXIT', _}, Value, _Op) ->
+    {binary, Value};
+maybe_convert_to_epoch(timestamp, Value, '>') ->
+    %% For strictly greater than, we increment any incomplete
+    %% date/time strings by one "unit", where the unit is the least
+    %% significant value provided by the user. "2016" becomes
+    %% "2017". "2016-07-05T11" becomes "2016-07-05T12".
+    CompleteFun = fun(DT) -> jam:expand(jam:increment(DT, 1), second) end,
+    {integer, do_epoch(Value, CompleteFun, 3)};
+maybe_convert_to_epoch(timestamp, Value, _Op) ->
+    CompleteFun = fun(DT) -> jam:expand(DT, second) end,
+    {integer, do_epoch(Value, CompleteFun, 3)};
+maybe_convert_to_epoch(_Type, Value, _Op) ->
+    {binary, Value}.
+
+%% I haven't investigated why the values are a nested list
+convert_insert_timestamps(Mod, Fields, [Values]) ->
```

Comment:
The third argument is a list of rows, each a list of values. As is, you are only taking a single row -- this will crash on statements inserting multiple rows (like "INSERT INTO t8 VALUES (1,7),(2,8),(3,9)").